pkg/collector/corechecks/cluster/kubernetes_state_label_joins.go

```diff
@@ -0,0 +1,208 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-2020 Datadog, Inc.
+
+// +build kubeapiserver
+
+package cluster
+
+import (
+	ksmstore "github.com/DataDog/datadog-agent/pkg/kubestatemetrics/store"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+/*
+   labelJoiner ingests all the metrics used by label joins to build a tree that can then be used to efficiently find which labels should be added to other metrics.
+
+   For ex., for the following JoinsConfig:
+
+     "kube_deployment_labels":
+       LabelsToMatch: ["namespace", "deployment"]
+       LabelsToGet:   ["chart_name", "chart_version"]
+
+   and the following metrics:
+
+     kube_deployment_labels {"namespace": "ns-a", "deployment": "foo", "chart_name": "foo", "chart_version": "1.0"}
+     kube_deployment_labels {"namespace": "ns-a", "deployment": "bar", "chart_name": "bar", "chart_version": "1.1"}
+     kube_deployment_labels {"namespace": "ns-b", "deployment": "foo", "chart_name": "foo", "chart_version": "2.0"}
+
+   it will create the following tree:
+
+   kube_deployment_labels
+     ├─ ns-a
```

Comment:
should we make sure the key is unique here? something like `namespace:ns-a`. It's an edge case yes but nothing prevents from having two labels to match with the same values. (we might need to add a test case for it)

Reply:
If there’s two metrics with the same values for the `labelsToMatch`, then, there will be a single node in the tree but the `labelsToAdd` of the leaf node will be the union of the `labelsToGet` of those two metrics.
I’ve added a test for this case.